app/src/modules/whyis-dataset.js

```diff
@@ -0,0 +1,279 @@
+    // Fields may have multiple values, so loop through all
+    for (const val in value) {
+      // type, value and id aren't in datasetFieldURIs dictionary
+      // but they are valid keys, so set the value to their value
+      if (['@type', '@value', '@id'].includes(val)) {
+        // if the value of val is an allowed field, use the field's value
+        // e.g., type = organization, and organization -> foaf:Organization
+        fieldDict[val] = Object.getOwnPropertyDescriptor(datasetFieldUris, value[val].toLowerCase())?.value ?? value[val]
+      } else if (Object.getOwnPropertyDescriptor(datasetFieldUris, val.toLowerCase())) { // Recursive case (val is an allowed field)
+        fieldDict[datasetFieldUris[val.toLowerCase()]] = recursiveFieldSetter([datasetFieldUris[val.toLowerCase()], value[val]])
+      } else {
+        fieldDict['@value'] = value
+      }
+    }
+    return fieldDict
+  }
+}
+
+// Blank dataset
+function getDefaultDataset () {
+  return Object.assign({}, defaultDataset)
+}
+
+function deleteDataset (datasetUri) {
+  return listNanopubs(datasetUri)
+    .then(nanopubs => {
+      Promise.all(nanopubs.map(nanopub => deleteNanopub(nanopub.np)))
+    }
+    )
+}
+
+// Handle all of the uploads as multipart form
+async function saveDataset (dataset, fileList, image, guuid) {
+  let p = Promise.resolve()
+  if (dataset.uri) {
+    p = deleteDataset(dataset.uri)
+  } else if (arguments.length === 1) {
+    dataset.uri = generateDatasetId()
+  } else {
+    dataset.uri = generateDatasetId(guuid)
+  }
+  const [distrRes, imgRes] = await Promise.all([
+    saveDatasetFiles(fileList),
+    saveDatasetFiles([image]), // TODO: verify we want to use the same function for depiction
+    p
+  ])
+  const datasetLd = buildDatasetLd(dataset)
+  if (distrRes?.files?.length) {
+    datasetLd[datasetFieldUris.distribution] = buildDistrLd(distrRes?.files)
+  } if (imgRes?.files?.length) {
+    datasetLd[datasetFieldUris.depiction] = buildDepictionLd(imgRes?.files?.[0], dataset.uri)
+  } return postNewNanopub(datasetLd)
+  // TODO: Error handling
+}
+
+async function saveDatasetFiles (fileList) {
+  if (fileList.length) {
+    const url = `${window.location.origin}/api/files/upload`
+    const formData = new FormData()
+    fileList.forEach((file) => formData.append('uploadfile', file?.file ?? file))
+    const result = await fetch(url, {
+      method: 'POST',
+      body: formData,
+      redirect: 'follow',
+      headers: {
+        Authorization: store.getters['auth/token']
+      }
+    })
+    return await result.json()
+    // TODO: Error handling
+  }
+}
+
+function buildDistrLd (fileList) {
+  const distrLDs = Array(fileList.length)
+  Array
+    .from(Array(fileList.length).keys())
+    .map(x => {
+      // TODO: check if we want to keep distribution uri as /explorer/dataset/id/filename and redirect for download
+      distrLDs[x] = {
+        '@id': `${window.location.origin}/api/files?filename=${fileList[x].filename}`,
```

Comment:
I think this should be `${window.location.origin}/api/files/get?filename=${fileList[x].filename}`